File: src/magpy/objects/mc_event.py
from dataclasses import dataclass
from enum import Enum
from typing import List

import jax.numpy as jnp

class MCEventIndices(Enum):
    TRUE_NEUTRINO_ENERGY = 0
    TRUE_Q2 = 1
    RECO_NEUTRINO_ENERGY = 2
    INTERACTION_MODE = 3
    START_NU = 4
    END_NU = 5
    
    TARGET = 6
    WEIGHT = 7
    NENTRIES = 8
    DUMMY = -1


@dataclass
class MCEvent:
    '''
    Represents a single event in the mc using JAX arrays
    '''
    true_neutrino_energy: float  # True neutrino energy
    true_q2: float  # True Q2
    reco_neutrino_energy: float  # Reconstructed neutrino energy
    interaction_mode: int  # Interaction mode
    target: int  # PDG for target
    
    start_nu: int  # PDG for incoming neutrino
    end_nu: int  # PDG for outgoing neutrino
    weight: float = 0 # Event weight
    
    # Now we want to make it tensor-y
    def to_array(self):
        return jnp.array([
            self.true_neutrino_energy,
            self.true_q2,
            self.reco_neutrino_energy,
            self.interaction_mode,
            self.start_nu,
            self.end_nu,
            self.target,
            self.weight,
        ], dtype=jnp.float64)


class MCEventMonolith:
    '''
    Monolith of MC Events
    '''
    def __init__(self, mc_event_list: List[MCEvent]):
        # Stack all events into a JAX array
        self._mc_event_monolith = jnp.stack(
            [event.to_array() for event in mc_event_list]
        )
        
        self._n_events = len(mc_event_list)

    @property
    def monolith(self):
        return self._mc_event_monolith

    @property
    def n_events(self):
        return self._n_events
    
    def __len__(self):
        return self._n_events


# For backward compatibility
# MCEvent = JAXMCEvent
# MCEventMonolith = JAXMCEventMonolith

File: cpp/src/mc_event.cpp
#include "mc_event.h"

namespace magpy {

MCEventMonolith::MCEventMonolith(const std::vector<MCEvent>& events) {
    const size_t n_events = events.size();
    const size_t n_cols = 8; // Number of event properties
    
    monolith_.resize(n_events, n_cols);
    
    for (size_t i = 0; i < n_events; ++i) {
        const auto& event = events[i];
        monolith_(i, static_cast<int>(MCEventIndices::TRUE_NEUTRINO_ENERGY)) = event.true_neutrino_energy;
        monolith_(i, static_cast<int>(MCEventIndices::TRUE_Q2)) = event.true_q2;
        monolith_(i, static_cast<int>(MCEventIndices::RECO_NEUTRINO_ENERGY)) = event.reco_neutrino_energy;
        monolith_(i, static_cast<int>(MCEventIndices::INTERACTION_MODE)) = event.interaction_mode;
        monolith_(i, static_cast<int>(MCEventIndices::START_NU)) = event.start_nu;
        monolith_(i, static_cast<int>(MCEventIndices::END_NU)) = event.end_nu;
        monolith_(i, static_cast<int>(MCEventIndices::TARGET)) = event.target;
        monolith_(i, static_cast<int>(MCEventIndices::WEIGHT)) = event.weight;
    }
}

} // namespace magpy
